twitter_scrooge/twitter_scrooge.bzl

```diff
@@ -281,99 +320,158 @@ def _scrooge_aspect_impl(target, ctx):
                 for d in ctx.rule.attr.deps
             ] + [target_ti],
         )
+    imps = [j[JavaInfo] for j in ctx.attr._implicit_compile_deps]
+
+    return (
+        target_ti,
+        transitive_ti,
+        deps,
+        imps,
+    )
+
+def _compile_thrift_to_language(target_ti, transitive_ti, language, target, ctx):
+    """Calls scrooge to compile thrift to the language specified in `language`.
+    Returns the name of the compiled jar."""
+
+    scrooge_file = ctx.actions.declare_file(
+        target.label.name + "_scrooge_{}.srcjar".format(language),
+    )
 
     # we sort so the inputs are always the same for caching
     compile_thrifts = sorted(target_ti.srcs.to_list())
-    imps = [j[JavaInfo] for j in ctx.attr._implicit_compile_deps]
-    if compile_thrifts:
-        # we sort so the inputs are always the same for caching
-        compile_thrift_map = {}
-        for ct in compile_thrifts:
-            compile_thrift_map[ct] = True
-        include_thrifts = sorted([
-            trans
-            for trans in transitive_ti.transitive_srcs.to_list()
-            if trans not in compile_thrift_map
-        ])
-        scrooge_file = ctx.actions.declare_file(
-            target.label.name + "_scrooge.srcjar",
-        )
-        _compile_to_scala(
-            ctx,
-            target.label,
-            compile_thrifts,
-            include_thrifts,
-            scrooge_file,
-        )
 
-        src_jars = depset([scrooge_file])
+    compile_thrift_map = {}
+    for ct in compile_thrifts:
+        compile_thrift_map[ct] = True
+    include_thrifts = sorted([
+        trans
+        for trans in transitive_ti.transitive_srcs.to_list()
+        if trans not in compile_thrift_map
+    ])
+
+    _generate_jvm_code(
+        ctx,
+        target.label,
+        compile_thrifts,
+        include_thrifts,
+        scrooge_file,
+        language,
+    )
+    return scrooge_file
+
+def _common_aspect_implementation(target, ctx, language, compiler_function):
+    """Aspect implementation to generate code from thrift files in a language of choice, and then compile it.
+    Takes in a `language` (either "java" or "scala") and a function to compile the generated sources.
+    """
+    allowed_languages = ["java", "scala"]
+    if language not in allowed_languages:
+        fail("Trying to compile thrift to language {}, when only {} are allowed".format(language, allowed_languages))
```

Comment:
We don't have any test for this and honestly we can't because it's a private function and on one can call it.
Why do we need this protection btw?

Reply:
We don't actually. I'll remove it.